README.md
<div align="center">

# Cryptic Resolver
[![Test](https://github.com/cryptic-resolver/cr.rb/workflows/Test-Behavior/badge.svg)](https://github.com/cryptic-resolver/cr.rb/actions/workflows/test.yml)
[![word-count](https://img.shields.io/badge/Keywords%20Inlcuded-1911-brightgreen)](#cryptic-sheets)
[![Gem Version](https://badge.fury.io/rb/cr.rb.svg)](https://rubygems.org/gems/cr.rb)

![screenshot](./images/screenshot.png)

</div>


This command line tool `cr` is used to **record and explain cryptic commands, acronyms(initialism), abbreviations and so forth** in daily life.
The effort is to study etymology and know of naming conventions.

Not only can it be used in the computer filed, but also you can use this to manage your own knowledge base easily.

- Currently we have **1911** keywords explained in all of our official ***cryptic_xxx*** dictionaries.

<br>

## Usage

**Install**

> $ gem install cr.rb

**Use it in the command line**

```bash
$ cr emacs
# -> Emacs: Edit macros
# ->
# ->   A feature-rich editor
# ->
# -> SEE ALSO Vim

$ cr -s pattern
# -> search words matched with pattern

$ cr -u
# -> update all dictionaries

$ cr -a https://github.com/ccmywish/common_zh-Hans.cr
# -> Add your own knowledge base!

$ cr -a ccmywish/CRuby-Source-Code-Dictionary
# -> Add a user's dict from Github by username and repo

$ cr -l
# All cryptic-resolver official dicts can be listed (local dicts listed also)

$ cr -a x86
# -> Add official dict 'cryptic-resolver/cryptic_x86', no need to specify the prefix!

$ cr -c
# -> word count

$ cr -h
# -> show help for more operations
```

<br>


## Why?

The aim of this project is to:

1. Make cryptic things clear
2. Help maintain your own personal knowledge base

rather than

- Record the use of a command, for this you can refer to [tldr], [cheat] and so on.

<br>


<a name="official-dictionaries"></a>
## Default dictionaries

You can override the default dicts by using a [config file](#config-file).

- [![cryptic_common](https://github.com/cryptic-resolver/cryptic_common/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_common/actions/workflows/test.yml) [cryptic_common]
- [![cryptic_computer](https://github.com/cryptic-resolver/cryptic_computer/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_computer/actions/workflows/test.yml) [cryptic_computer]
- [![cryptic_windows](https://github.com/cryptic-resolver/cryptic_windows/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_windows/actions/workflows/test.yml) [cryptic_windows]
- [![cryptic_linux](https://github.com/cryptic-resolver/cryptic_linux/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_linux/actions/workflows/test.yml) [cryptic_linux]
- [![cryptic_technology](https://github.com/cryptic-resolver/cryptic_technology/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_technology/actions/workflows/test.yml) [cryptic_technology]

<br>

## Other official dictionaries

```bash
# List all official dicts (and local dicts)
$ cr -l

# Easy to add!
$ cr -a economy

$ cr -a medicine

$ cr -a electronics

...
```

- [![cryptic_economy](https://github.com/cryptic-resolver/cryptic_economy/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_economy/actions/workflows/test.yml) [cryptic_economy]
- [![cryptic_medicine](https://github.com/cryptic-resolver/cryptic_medicine/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_medicine/actions/workflows/test.yml) [cryptic_medicine]
- [![cryptic_electronics](https://github.com/cryptic-resolver/cryptic_electronics/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_electronics/actions/workflows/test.yml) [cryptic_electronics]
- [![cryptic_science](https://github.com/cryptic-resolver/cryptic_science/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_science/actions/workflows/test.yml) [cryptic_science]
- [![cryptic_math](https://github.com/cryptic-resolver/cryptic_math/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_math/actions/workflows/test.yml) [cryptic_math]

<br>

**Specific dicts:**

- [![cryptic_x86](https://github.com/cryptic-resolver/cryptic_x86/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_x86/actions/workflows/test.yml) [cryptic_x86]
- [![cryptic_signal](https://github.com/cryptic-resolver/cryptic_signal/workflows/Test-Dict/badge.svg)](https://github.com/cryptic-resolver/cryptic_signal/actions/workflows/test.yml) [cryptic_signal]

<br>

**Feature dicts:**

- [ccmywish/CRuby-Source-Code-Dictionary](https://github.com/ccmywish/CRuby-Source-Code-Dictionary)
- [ccmywish/mruby-Source-Code-Dictionary](https://github.com/ccmywish/mruby-Source-Code-Dictionary)

<br>

## Implementation

`cr` is written in **Ruby**. You can implement this tool in any other language you like (name your projects as `cr_Python` or `cr.py` for example), just remember to reuse [cryptic_computer] and other dictionaries which are the core parts anyone can contribute to. See [doc/dev.md](doc/dev.md) to help develop && maintain this tool.

<br>

## Library layout

A library is a folder which contain many dictionaries (which are just `git` repos). By default, dictionaries will be downloaded in OS-specific folder (Using [`StandardPath.app_data`](https://github.com/ccmywish/standard_path)), this can't be configured. But, the user can specify an extra library using a config file.

```
Library
.
├── cryptic_common
├── cryptic_computer
├── cryptic_linux
├── cryptic_windows
├── ...
├── cryptic_economy
├── cryptic_x86
└── user_exam_dictionary
```

<br>

<a name="config-file"></a>
## Config file

You can use a config file (in `TOML` format) which is specified by the environment variable `CRYPTIC_RESOLVER_CONFIG`. You can control two options in this config file:

1. Override the default dicts
2. Add an extra library for yourself, this way, you can manage multiple knowledge bases of your own, without the hassle of maintaining multiple git repos.

```toml
# cryptic-resolver.toml example

DEFAULT_DICTS = [
  "https://github.com/cryptic-resolver/cryptic_x86",
  "https://github.com/cryptic-resolver/cryptic_electronics"
]

EXTRA_LIBRARY = "C:\\MyData\\cr"
```

<br>

## Dictionary layout

`Dictionary` is a knowledge base. Every dictionary should be a `git` repository, and each consists of many files (we call these `sheets`):
```
Dictionary
.
├── 0-9.toml
├── a.toml
├── b.toml
├── c.toml
├── ...
├── y.toml
└── z.toml
```

## Sheet format (File format)

In every file(or sheet), your definition format looks like this in pure **toml**:
```toml
# A normal definition
#
# NOTICE:
#   You MUST keep the key in lower case.
#   Use a key 'name' to display its original form
#   Because the case sometimes contains details to help we understand
#
#   And 'name' && 'desc' is both MUST-HAVE.
#   But if you use 'same', all other infos are not needed.
#
[xdg]
name = "XDG"
desc = "Cross Desktop Group"

# If you want to explain more, use 'more'
[xxd]
name = "xxd"
desc = "hex file dump"
more = "Why call this 'xxd' rather than 'xd'?? Maybe a historical reason"

[xdm]
name = "XDM"
desc = "xdm without specifier"

# You can add a subkey as a category specifier to differ
# While the main word must be lower case, the category name can be any case!
[xdm.Download]
name = "XDM"
desc = "eXtreme Download Manager"

[xdm.Display]
name = "XDM"
desc = "X Display Manager"
```

**The search strategy is to treat words with or without a category specifier in the same way.**


More features:
```toml
[jpeg]
name = "JPEG"
desc = "Joint Photographic Experts Group"
more = "Introduced in 1992. A commonly used method of lossy compression for digital images"
see = 'MPG' # This is a `see also`
# see = ['MPG', 'PNG']  # when multiple see also, use array

[jpg]
name = "JPG"  # Show a good name
same = "jpeg" # We just need to redirect this. No duplicate!

[sth]
# You must point to a exact position, not just xdm
# And we use '=>' symbol to split it into two parts
#   the first will output to user
#   the second is for internal jump
same = "Extreme Downloader =>xdm.Download"


# The 'dot' keyword supported using quoted strings
["h.265"]
name = "H.265"
desc = "A video compression standard"

```

## Name collision

In one sheet, you should consider adding a category specifier to differ each other like the example above.

*But what if a dictionary has 'gdm' while another also has a 'GDM'?*

> cr can handle this.

*But what if a sheet has two 'gdm'?*

> This will lead to toml's parser library fail. You have these solutions
> 1. Use a better lint for example: [VSCode's Even Better TOML](https://github.com/tamasfe/taplo)
> 2. Watch the fail message, you may notice 'override path xxx', the xxx is the collision, you may correct it back manually.

<br>

## LICENSE

`cr.rb` itself is under MIT

[Official dictionaries](#official-dictionaries) are all under CC-BY-4.0

[cryptic_common]: https://github.com/cryptic-resolver/cryptic_common
[cryptic_computer]: https://github.com/cryptic-resolver/cryptic_computer
[cryptic_windows]: https://github.com/cryptic-resolver/cryptic_windows
[cryptic_linux]: https://github.com/cryptic-resolver/cryptic_linux
[cryptic_technology]: https://github.com/cryptic-resolver/cryptic_technology

[tldr]: https://github.com/tldr-pages/tldr
[cheat]: https://github.com/cheat/cheat

[cryptic_economy]: https://github.com/cryptic-resolver/cryptic_economy
[cryptic_medicine]: https://github.com/cryptic-resolver/cryptic_medicine
[cryptic_electronics]: https://github.com/cryptic-resolver/cryptic_electronics
[cryptic_science]: https://github.com/cryptic-resolver/cryptic_science
[cryptic_math]: https://github.com/cryptic-resolver/cryptic_math

[cryptic_x86]: https://github.com/cryptic-resolver/cryptic_x86
[cryptic_signal]: https://github.com/cryptic-resolver/cryptic_signal


CHANGELOG.md
# Changelog

## [Unreleased](#) (2023-08-15)

<br>

## [v4.1.8](#) (2023-08-15)

### Bug fixes:

- Fix the path that has spaces when `git clone`

<br>

## [v4.1.7](#) (2023-07-29)

### Deprecations:

- Change default lib location

<br>

## [v4.1.6](#) (2023-05-15)

### Enhancements:

- Rename some internal APIs
- Using `rainbow` rather than `colorator`
- Using `minitest` rather than `test-unit`

<br>

## [v4.1.5](#) (2023-05-09)

### Enhancements:

- Compact not found info
- Refactor lib structure
- Use `colorator` rather than handmade lib

<br>

## [v4.1.4](#) (2023-03-22)

### Bug fixes:

- Fix bug of `-l` list file: [GitHub issue 8](https://github.com/cryptic-resolver/cr.rb/issues/8)

### Deprecations:

- Remove `cryptic_dos` and `cryptic_mechanical` dictionaries

<br>

## [v4.1.3](#) (2023-02-13)

### Bug fixes:

- Fix bug: read unrelated files in dictionary

<br>

## [v4.1.2](#) (2023-02-13)

### Bug fixes:

- Fix bug: `cr -c` counts twice

<br>

## [v4.1.1](#) (2023-02-13)

### Bug fixes:

- Fix `require` error in `lib/cr.rb`

<br>

## [v4.1.0](#) (2023-02-13)

### New features:

- Add new feature dictionary: mruby source code dictionary

### Enhancements:

- Huge refactor

<br>

## [v4.0.3](#) (2023-02-05)

### Enhancements:

- Remove command line see also print trailing space
- Colorize `NOT FOUND` info

### Bug fixes:

- Fix bug of `$cr` help messages, see [commit 5d8399c](https://github.com/cryptic-resolver/cr.rb/commit/5d8399c6a523001478202a90c90c81f76bd694cb)

<br>

## [v4.0.2](#) (2022-11-29)

### Enhancements:

- Use `WARN` to make CI test more accurate

<br>

## [v4.0.1](#) (2022-11-28)

### Bug fixes:

- Fix word count in adding new dict and pulling default dicts

<br>

## [v4.0.0](#) (2022-11-27)

### Enhancements:

- Use new index 0-9.toml
- Use new jump symbol `=>`
- Add support for user config
- List official dicts
- Word count different libraries

### Deprecations:

- Not use old index 0123456789.toml
- Not use old jump symbol `<=>`

<br>

## [v3.21.0](#) (2022-10-30)

### Enhancements:

- Add `cryptic_linux` to default dicts

<br>

## [v3.20.0](#) (2022-10-20)

### Enhancements:

- Update tests
- CI updates all cryptic_xxx dicts words
- `bin/wc` counts all local dicts
- Choose different default dicts

<br>

## [v3.19.0](#) (2022-04-15)

### New features:

- Support see_also can be one word
- Simplify adding dictionary with `user/repo`

<br>

## [v3.18.2](#) (2022-04-15)

### Enhancements:

- Add CI test

### Bug fixes:

- Ensure cr home dir exists when adding new dict

<br>

## [v3.18.0](#) (2022-04-11)

### Deprecations:

- Use `more` rather than `full`
- Use `name` rather than `disp`

<br>

## [v3.17.0](#) (2022-04-11)

### New features:

- Support to show the word's own better name when jumping

### Enhancements:

- Add `cr.rb` gem
- Add tests

<br>

## [v3.16](#) (2022-04-05)

### Enhancements:

- Always add default dicts when help

### Bug fixes:

- Fix error on `cr -u` when first installed, this is because of double word counting
- Fix no dir error when first installed, this is because of the word count of help

<br>

## [v3.15](#) (2022-04-01)

### Bug fixes:

- Fix user dict word count

<br>

## [v3.14](#) (2022-03-24)

### Enhancements:

- Either `desc` or `more` exists is OK

<br>

## [v3.13](#) (2022-03-24)

### New features:

- Allow double or more jumps

<br>

## [v3.12](#) (2022-03-24)

### Enhancements:

- add dependency to print

<br>

## [v3.11](#) (2022-03-22)

### New features:

- Add support to search words via regexp

### Enhancements:

- Add support to jump to dotted word

<br>

## [v3.10](#) (2022-03-18)

### New features:

- Add spaceship delimiter for jump info

### Enhancements:

- Migrate from **github_changelog_generator** to **chlog**
- Yank version `v4,5,6` from RubyGems, release on `v3.x` series

<br>

## [v3.0 - v3.9](#)

### New features:

- Add parallel update and pull support
- Add `-l`, `-d` to manage local dictionaries

### Bug fixes:

- Fix many small bugs

<br>

## [v1.0 - v2.3](#)

### New features:

- Support dotted words [GitHub issue #5](https://github.com/cryptic-resolver/cr/issues/5)
- Warn to incomplete explanation [GitHub issue #4](https://github.com/cryptic-resolver/cr/issues/4)
- Add original keyword to print [GitHub issue #3](https://github.com/cryptic-resolver/cr/issues/3)
- Add 'see' keyword [GitHub issue #2](https://github.com/cryptic-resolver/cr/issues/2)
- Add 'same' keyword  [GitHub issue #1](https://github.com/cryptic-resolver/cr/issues/1)

<br>

## [Initialize](#) (2021-07-08)

<br>

<hr>

This Changelog is maintained with [chlog](https://github.com/ccmywish/chlog)



test/basic.rb
def cr(arg)
  
  if test 'd', 'bin'
    `ruby ./bin/cr #{arg}`
  else
    `ruby ../bin/cr #{arg}`
  end
  
end

a =  cr("emacs")

b = <<~"EOC"
\e[32mFrom: cryptic_computer\e[0m

  Emacs: Edit macros

  a feature-rich editor

\e[35mSEE ALSO \e[0m\e[4mVim\e[0m 

EOC

# notice: only one line left to EOC

puts a == b


test/gen_output.rb
require_relative 'basic.rb'

commands = %w[
  emacs
  jpg
  loc
  bcd
  gles
  eg.
  eg
  xdm
  ide
  mri
  smt
  gdm
]

output = ""

commands.each do
  output += "-----------#{_1}-------------\n"
  output += cr _1
end

# convert the ESC to real \e
output.gsub!("\e", '\e')

File.write('gen_output.txt', output)


lib/cr.rb
# ------------------------------------------------------
# File          : cr.rb
# Authors       : Aoran Zeng <ccmywish@qq.com>
# Created on    : <2022-04-15>
# Last modified : <2023-08-15>
#
# cr:
#
#   This file is the lib of `cr.rb``
# ------------------------------------------------------

require 'rainbow/refinement'
require_relative 'cr/version'

class CrypticResolver::Resolver

  using Rainbow

  require_relative 'cr/counter'

  # Notice that we only update the Default library, not Extra library
  def update_dicts

    @counter.count_def_lib(display: false)
    old_wc = @counter.word_count_of_def_lib

    puts "cr: Updating all dicts in Default library..."

    begin
    Dir.chdir DEFAULT_LIB_PATH do
      if Gem.win_platform?
        # Windows doesn't have fork
        Dir.children(DEFAULT_LIB_PATH).each do |dict|
          next if File.file? dict
          puts "cr: Wait to update #{dict}..."
          `git -C ./#{dict} pull -q`
        end
      else
        # *nix
        Dir.children(DEFAULT_LIB_PATH).each do |dict|
          next if File.file? dict
          fork do
            puts "cr: Wait to update #{dict}..."
          `git -C ./#{dict} pull -q`
          end
        end
        Process.waitall
      end # end if/else
    end

    rescue Interrupt
    abort "cr: Cancel update"
    end

    puts "cr: Update done"

    # clear
    @counter.word_count_of_def_lib = 0
    # recount
    @counter.count_def_lib(display: false)

    new_wc = @counter.word_count_of_def_lib
    puts ; puts "#{new_wc - old_wc} words added in Default library"
  end


  def add_dict(repo)
    if repo.nil?
      abort "cr: Need an argument!".bold.red
    end

    # Simplify adding dictionary
    if !repo.start_with?("https://") and !repo.start_with?("git@")
      if repo.include?('/')
        repo = "https://github.com/#{repo}.git"
      else
        repo = "https://github.com/cryptic-resolver/cryptic_#{repo}.git"
      end
    end

    begin
    puts "cr: Adding new dictionary..."
    # Note that, we must add "" to surround the dir
    # Because some path (e.g. macOS) will have spaces
    `git -C "#{DEFAULT_LIB_PATH}" clone #{repo} -q`
    rescue Interrupt
    abort "cr: Cancel add dict"
    end

    puts "cr: Add new dictionary done"

    # github/com/ccmywish/ruby_knowledge(.git)
    dict = repo.split('/')[-1].delete_suffix('.git')
    wc = @counter.count_dict_words(DEFAULT_LIB_PATH ,dict)
    puts ; puts "#{wc} words added"
  end


  def del_dict(repo)
    if repo.nil?
      abort "cr: Need an argument!".bold.red
    end
    Dir.chdir DEFAULT_LIB_PATH do
      begin
      # Dir.rmdir repo        # Can't rm a filled dir
      # FileUtils.rmdir repo  # Can't rm a filled dir
      FileUtils.rm_rf repo
      puts "cr: Delete dictionary #{repo.bold.green} done"
      rescue Exception => e
      puts "cr: #{e}".bold.red
      list_dicts
      end
    end
  end


  def load_sheet(library, dict, sheet_name)
    file = library + "/#{dict}/#{sheet_name}.toml"
    if File.exist? file
      return Tomlrb.load_file file # gem 'tomlrb'
      # return TOML.load_file file # gem 'toml'
    else nil end
  end


=begin
Pretty print the info of the given word

A info looks like this
  emacs = {
    name = "Emacs"
    desc = "edit macros"
    more = "a feature-rich editor"
    see  = ["Vim"]
  }

@param info [Hash] the information of the given word (mapped to a keyword in TOML)
=end
  def pp_info(info)
    name = info['name'] || "No name!".red  # keyword `or` is invalid here in Ruby

    desc = info['desc']
    more = info['more']

    if desc
      puts "\n  #{name}: #{desc}"
      print "\n  ",more,"\n" if more
    else
      puts "\n  #{name}"
      print "\n  ",more,"\n" if more
    end

    if see_also = info['see']
      print "\n", "SEE ALSO ".magenta
      if see_also.is_a?(Array)
        last_ndx = see_also.size - 1
        see_also.each_with_index do |x,i|
          if last_ndx == i
            print x.underline  # Last word doesn't show space
          else
            print x.underline, ' '
          end
        end
      else
        print see_also.underline
      end
      puts
    end
    puts
  end


  # Print default cryptic_ dicts
  def pp_dict(dict)
    puts "From: #{dict}".green
  end


=begin
  Used for synonym jump
  Because we absolutely jump to a must-have word
  So we can directly lookup to it

  Notice that, we must jump to a specific word definition
  So in the toml file, you must specify the precise word.
  If it has multiple meanings, for example

  [blah]
  same = "xdg"  # this is wrong, because xdg has multiple
                # definitions, and all of them specify a
                # category

  [blah]
  same = "XDG downloader =>xdg.Download" # this is correct

  [blah]
  name = "BlaH" # You may want to display a better name first
  same = "XDG downloader =>xdg.Download" # this is correct
=end
  def pp_same_info(library, dict, word, cache_content, same_key, own_name)

    # If it's a synonym for anther word,
    # we should lookup into this dict again, but maybe with a different file

    # file name
    x = word.chr.downcase

=begin
    dictionary maintainer must obey the rule for xxx.yyy word:
    xxx should be lower case
    yyy can be any case

    Because yyy should clearly explain the category info, IBM is better than ibm
    Implementation should not be too simple if we want to stress the function we
    expect.

    'jump to' will output to user, so this is important not only inside our sheet.

      same = "XDM downloader=>xdm.Download"

    We split 'same' key into two parts via spaceship symbol `=>`, first part will
    output to user, the second part is for internal jump.
=end
    jump_to, same = same_key.split("=>")
    same =  jump_to if same.nil?

    unless own_name
      own_name = word
    end
    puts own_name.bold.blue + ' redirects to ' + jump_to.bold.blue

=begin
    As '.' is used to delimit a word and a category, what if
    we jump to a dotted word?

      [eg]
      same = "e.g." # this must lead to a wrong resolution to
                    # word 'e', category 'g'

      All you need is to be like this:

      [eg]
      same = "'e.g.'" # cr will notice the single quote
=end
    if same =~ /^'(.*)'$/
      same, category = $1, nil
    else
      same, category = same.split('.')
    end

    if same.chr == x
      # No need to load another dictionary if match
      sheet_content = cache_content
    else
      sheet_content = load_sheet(library, dict, same.chr.downcase)
    end

    if category.nil?
      info = sheet_content[same]
    else
      info = sheet_content[same][category]
    end

    if info.nil?
      puts "WARN: Synonym jumps to the wrong place `#{same}`,
      Please consider fixing this in `#{x}.toml` of the dictionary `#{dict}`".red
      # exit
      return false
    # double or more jumps
    elsif same_key = info['same']
      own_name = info['name']
      return pp_same_info(library, dict, same, cache_content, same_key, own_name)
    else
      pp_info(info)
      return true
    end
  end


  # Lookup the given word in a sheet (a toml file) and also print.
  # The core idea is that:
  #
  # 1. if the word is `same` with another synonym, it will directly jump to
  #   a word in this dictionary, but maybe a different sheet.
  #
  # 2. load the toml file and check the given word
  #   2.1 with no category specifier
  #       [abcd]
  #   2.2 with category specifier
  #       [abcd.tYPe]
  #
  def lookup(library, dict, file, word)
    sheet_content = load_sheet(library, dict, file)
    return false if sheet_content.nil?

    info = sheet_content[word]
    return false if info.nil?

    # Warn if the info is empty. For example:
    #   emacs = { }
    if info.size == 0
      abort "WARN: Lack of everything of the given word
      Please consider fixing this in the dict `#{dict}`".red
    end

    # Word with no category specifier
    # We call this meaning as type 1
    type_1_exist_flag = false

    # if already jump, don't check the word itself
    is_jump = false

    # synonym info print
    if same_key = info['same']
      own_name = info['name']
      pp_dict(dict)
      pp_same_info(library, dict, word, sheet_content, same_key, own_name)
      # It's also a type 1
      type_1_exist_flag = true
      is_jump = true
    end

    # normal info print
    # To developer:
    #   The word should at least has one of `desc` and `more`
    #   But when none exists, this may not be considered wrong,
    #   Because the type2 make the case too.
    #
    #   So, just ignore it, even if it's really a mistake(insignificant)
    #   by dictionary maintainers.
    #
    if !is_jump && (info.has_key?('desc') || info.has_key?('more'))
      pp_dict(dict)
      pp_info(info)
      type_1_exist_flag = true
    end

    # Meanings with category specifier
    # We call this meaning as type 2
    categories = info.keys - ["name", "desc", "more", "same", "see"]

    if !categories.empty?

      if type_1_exist_flag
        print "OR".bold.blue, "\n"
      else
        pp_dict(dict)
      end

      categories.each do |meaning|
        info0 = sheet_content[word][meaning]
        if same_key = info0['same']
          own_name = info0['name']
          pp_same_info(library, dict, word, sheet_content, same_key, own_name)
        else
          pp_info(info0)
        end

        # last meaning doesn't show this separate line
        print "OR".bold.blue, "\n" unless categories.last == meaning
      end
      return true
    elsif type_1_exist_flag then return true
    else return false end
  end


  # The main procedure of `cr`
  #
  #   1. Search the default library first
  #   2. Search the extra library if it does exist
  #
  # The `search` is done via the `lookup` function. It will print
  # the info while finding. If `lookup` always return false then
  # means lacking of this word in our dicts. So a welcomed
  # contribution is printed on the screen.
  #
  def resolve_word(word)

    word = word.downcase # downcase! would lead to frozen error in Ruby 2.7.2
    # The index is the toml file we'll look into
    index = word.chr
    case index
    when '0'..'9'
      index = '0-9'
    end

    # cache lookup's results
    results = []

    # First consider the default library
    default = Dir.children(DEFAULT_LIB_PATH)
    default.each do |dict|
      results << lookup(DEFAULT_LIB_PATH,dict,index,word)
    end

    # Then is the extra library
    if @extra_lib_path
      extra = Dir.children(@extra_lib_path)
      extra.each do |dict|
        results << lookup(@extra_lib_path,dict,index,word)
      end
    end

    unless results.include? true
      puts <<~NotFound
      cr: Not found anything about '#{word}'. You could try

      #{"case 1: Update all dictionaries".blue}
       #{"$ cr -u".yellow}
      #{"case 2: List available official and feature dictionaries".blue}
       #{"$ cr -l".yellow}
       #{"$ cr -a repo".yellow} (Add a specific dict to default lib)
      #{"case 3: Contribute to theses dictionaries".blue}
       Visit: https://github.com/cryptic-resolver

      NotFound
    else return end
  end


  # Delegate to `search_word_internal`
  #
  def search_related_words(pattern)
    found_or_not1 = false
    found_or_not2 = false

    found_or_not1 = search_related_words_internal(pattern, DEFAULT_LIB_PATH)
    if @extra_lib_path
      found_or_not2 = search_related_words_internal(pattern, @extra_lib_path)
    end

    if (found_or_not1 == false) && (found_or_not2 == false)
      puts "cr: No words match with #{pattern.inspect}".red ; puts
    end
  end


  # This routine is quite like `resolve_word`
  #
  # Notice:
  #   We handle two cases
  #
  #   1. the 'pattern' is the regexp itself
  #   2. the 'pattern' is like '/blah/'
  #
  #  The second is what Ruby and Perl users like to do, handle it!
  #
  def search_related_words_internal(pattern, library)

    if pattern.nil?
      abort "cr: Need an argument!".bold.red
    end

    if pattern =~ /^\/(.*)\/$/
      regexp = %r[#$1]
    else
      regexp = %r[#{pattern}]
    end

    found_or_not = false

    # Try to match every word in all dicts
    Dir.children(library).each do |dict|

      path = File.join(library, dict)
      next if File.file? path
      sheets = Dir.children(path).select do
        _1.end_with?('.toml')
      end

      similar_words_in_a_dict = []

      sheets.each do |sheet|
        sheet_content = load_sheet(library, dict, File.basename(sheet,'.toml'))

        sheet_content.keys.each do
          if _1 =~ regexp
            found_or_not = true
            similar_words_in_a_dict << _1
          end
        end
      # end of each sheet in a dict
      end

      unless similar_words_in_a_dict.empty?
        pp_dict(dict)
        require 'ls_table'
        LsTable.ls(similar_words_in_a_dict) do |e|
          puts e.blue
        end
        puts
      end
    end
    return found_or_not
  end


  # 1. List Default library's dicts
  # 2. List Extra   library's dicts
  # 3. List official dicts
  def list_dicts

    def _do_the_same_thing
      count = 0
      Dir.children(".").each do |dict|
        next if File.file? dict
        count += 1
        prefix = (count.to_s + '.').ljust 4
        puts "  #{prefix}#{dict}"
      end
    end

    Dir.chdir DEFAULT_LIB_PATH do
      puts "Default library: #{DEFAULT_LIB_PATH}".bold.green
      _do_the_same_thing
    end

    if @extra_lib_path
      puts
      Dir.chdir @extra_lib_path do
        puts "Extra library: #{@extra_lib_path}".bold.green
        _do_the_same_thing
      end
    end

    puts ; puts "Official dictionaries: (Add it by 'cr -a xxx')".bold.green
    puts RECOMMENDED_DICTS
  end


  def count_words
    @counter.count!(display: true)
  end

end



class CrypticResolver::Resolver

  using Rainbow

  require 'tomlrb'
  require 'fileutils'
  require 'standard_path'

  # attr_accessor   :default_dicts  # Default dictionaries lists

  DEFAULT_LIB_PATH = StandardPath.app_data 'Cryptic-Resolver'

  ORIGINAL_DEFAULT_DICTS = [
    "https://github.com/cryptic-resolver/cryptic_common.git",
    "https://github.com/cryptic-resolver/cryptic_computer.git",
    "https://github.com/cryptic-resolver/cryptic_windows.git",
    "https://github.com/cryptic-resolver/cryptic_linux.git",
    "https://github.com/cryptic-resolver/cryptic_technology"
  ]

  RECOMMENDED_DICTS = <<~EOF
  #{"Default:".yellow}
    common       computer    windows
    linux       technology

  #{"Field:".yellow}
    economy     medicine   electronics
    science       math

  #{"Specific:".yellow}
      x86        signal

  #{"Feature:".yellow}
    ccmywish/CRuby-Source-Code-Dictionary

  EOF



  attr_accessor :def_dicts,        # default dictionaries lists
                :extra_lib_path,   # Extra library path
                :counter           # word counter

  def initialize
    @counter =  Counter.new(self)

    # if user doesn't specify, we use the hard-coded defaults
    @def_dicts = ORIGINAL_DEFAULT_DICTS

    # The config file will override the default dicts, but not default library!
    if file = ENV['CRYPTIC_RESOLVER_CONFIG']

      if !test('f', file)
        abort "FATAL: Your CRYPTIC_RESOLVER_CONFIG is NOT a file!"
      end

      config = Tomlrb.load_file file

      if ret = config['DEFAULT_DICTS']
        @def_dicts = ret
      end

      @extra_lib_path ||= config['EXTRA_LIBRARY']
      # Prevent runtime error
      if ! test('d', @extra_lib_path)
        abort "FATAL: Your CRYPTIC_RESOLVER_CONFIG's option 'EXTRA_LIBRARY' is NOT a legal directory!"
      end
    else
      @extra_lib_path = nil
    end

    # Same with the pulled repo dirs' names in DEFAULT_LIB_PATH
    # @def_dicts_names = @def_dicts.map do |e|
    #  e.split('/').last.split('.').first
    # end


  end


  def is_there_any_dict?
    # Ensure the cr home dir exists
    FileUtils.mkdir_p(DEFAULT_LIB_PATH)
    #unless Dir.exist? DEFAULT_LIB_PATH
    # Dir.mkdir DEFAULT_LIB_PATH
    #end
    !Dir.empty? DEFAULT_LIB_PATH
  end


  def add_default_dicts_if_none_exists
    unless is_there_any_dict?
      puts "cr: Adding default dicts..."

      # This is used to display what you are pulling when adding dicts
      dicts_user_and_names = @def_dicts.map do |e|
        user, repo = e.split('/').last(2)
        repo = repo.split('.').first
        user + '/' + repo
      end

      begin
      if Gem.win_platform?
        # Windows doesn't have fork
        dicts_user_and_names.each_with_index do |name, i|
          puts "cr: Pulling #{name}..."
          `git -C "#{DEFAULT_LIB_PATH}" clone #{@def_dicts[i]} -q`
        end
      else
        # *nix-like
        dicts_user_and_names.each_with_index do |name, i|
          fork do
            puts "cr: Pulling #{name}..."
            `git -C "#{DEFAULT_LIB_PATH}" clone #{@def_dicts[i]} -q`
          end
        end
        Process.waitall
      end

      rescue Interrupt
      abort "cr: Cancel add default dicts"
      end

      puts "cr: Add done" ;
      @counter.count_def_lib(display: false)
      puts ; puts "#{@counter.word_count_of_def_lib} words added"
      @counter.reset!

      # Really added
      return true
    end
    # Not added
    return false
  end

end
